packages/realm/src/Realm.ts

```diff
@@ -284,24 +290,75 @@ export class Realm {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  static setLogLevel(_: LogLevel | LogArgs) {
+    // It is not possible to overload a static function: https://github.com/microsoft/TypeScript/issues/18945
+
+    const setLevel = (category: LogCategory, level: LogLevel) => {
+      assert(Object.values(LogCategory).includes(category));
+      const ref = binding.LogCategoryRef;
+      const c = ref.getCategory(category);
+      c.setDefaultLevelThreshold(toBindingLoggerLevel(level));
+    };
+
+    // FIXME: don't use `arguments` but find a proper type
+    if (arguments.length === 1) {
+      // eslint-disable-next-line prefer-rest-params
+      const arg = arguments[0];
+      if (arg.level) {
+        const level = arg.level as LogLevel;
+        if (arg.category) {
+          const category = arg.category as LogCategory;
+          setLevel(category, level);
+        } else {
+          Object.values(LogCategory).forEach((category) => {
+            setLevel(category, level);
+          });
+        }
+      } else {
+        const level = arg as LogLevel;
+        Object.values(LogCategory).forEach((category) => {
+          setLevel(category, level);
+        });
+      }
+    } else {
+      throw new Error(`Wrong number of arguments - expected 1, got ${arguments.length}`);
+    }
   }
 
   /**
    * Sets the logger callback.
    * @param loggerCallback - The callback invoked by the logger. The default callback uses `console.log`, `console.warn` and `console.error`, depending on the level of the message.
-   * @note The logger callback needs to be setup before opening the first realm.
+   * @note The logger callback needs to be setup before opening the first Realm.
    * @since 12.0.0
+   * @example
+   * Realm.setLogger(({ category, level, message }) => {
+   *   console.log(`[${category} - ${level}] ${message}`);
+   * });
    */
   static setLogger(loggerCallback: LoggerCallback) {
-    const logger = binding.Helpers.makeLogger((level, message) => {
-      loggerCallback(fromBindingLoggerLevelToLogLevel(level), message);
-    });
+    let logger: binding.Logger;
+
+    // This is a hack to check which of the two logger callbacks which are used
+    // It only works as the two callback type have different number of arguments, and it will
+    // probably produce odd error messages if the logger is set by `setLogger((...args) => console.log(args))`.
+    if (loggerCallback.length === 2) {
+      const cb = loggerCallback as LoggerCallback1;
+      logger = binding.Helpers.makeLogger((_category, level, message) => {
+        cb(fromBindingLoggerLevelToLogLevel(level), message);
+      });
+    } else {
+      const cb = loggerCallback as LoggerCallback2;
+      logger = binding.Helpers.makeLogger((category, level, message) => {
+        cb({ category: category as LogCategory, level: fromBindingLoggerLevelToLogLevel(level), message });
+      });
+    }
     binding.Logger.setDefaultLogger(logger);
```

Comment:
Nice check on the function's length 🙂

To prevent the misuse of `(...args)` as you mentioned, and to remove the need for `as LoggerCallback1/2`, we could throw for the wrong number of callback args and also extract the if-conditions into two functions in e.g. `Logger.ts` (with whatever names you think are suitable):

```ts
export function isLoggerWithLevel(logger: unknown): logger is LoggerCallback1 {
  return typeof logger === "function" && logger.length === 2;
}

export function isLoggerWithOptions(logger: unknown): logger is LoggerCallback2 {
  return typeof logger === "function" && logger.length === 1;
}
```

Then the `setLogger()` function could utilize those:

```ts
  static setLogger(loggerCallback: LoggerCallback) {
    let logger: binding.Logger;

    if (isLoggerWithLevel(loggerCallback)) {
      logger = binding.Helpers.makeLogger((_, level, message) => {
        loggerCallback(fromBindingLoggerLevelToLogLevel(level), message);
      });
    } else if (isLoggerWithOptions(loggerCallback)) {
      logger = binding.Helpers.makeLogger((category, level, message) => {
        loggerCallback({
          category: category as LogCategory,
          level: fromBindingLoggerLevelToLogLevel(level),
          message,
        });
      });
    } else {
      throw new Error(`Unexpected logger passed.`);
    }

    binding.Logger.setDefaultLogger(logger);
  }
```

Reply:
I like the suggestion to use of an assert function, but it would be even greater if it used a pure function (`loggerFromCallback` or something) returning the `logger` instead of a local mutable variable 😉